### DISCUSSION TITLE
검증전략 관련해서 생각해볼만한 것들

### DISCUSSION THREAD

**Discussion by MoonJaeHoon** [bcaitech1/p4-dkt-freshtomato#47] [Tip]
## 검증전략 (특히, Train Valid Split)
서일 님의 현재 1등 [토론 공유글 ](http://boostcamp.stages.ai/competitions/42/discussion/post/357)중에 검증전략 부분을 정리해보았음.. 바빠서 노트북 내부까지 볼 시간 없는 사람은
지금 이 글 한번 스윽 보면 인사이트 생길 수 있고 좋을듯~!!

### 간단요약
- 수많은 row의 Train을 꼭 다써야 할까?
- Valid도 `사용자의 마지막 문항` 뿐만 아니라 특별하게 구축해볼 수는 없을까?

먼저, 우리가 일반적으로 검증전략으로서 쓰고있던 코드들을 살펴보면 다음과 같다.
```python
size = int(len(data) * ratio)
data_1 = data[:size]
data_2 = data[size:]
```
혹은
```python
for train_idx, test_idx in skf.split(train_all_data, user_stratified_key):
    args.fold = fold_counter
  
    train_data = train_all_data[train_idx]
    valid_data = train_all_data[test_idx]
```
(여기서 train_idx와 test_idx는 user_id를 나눈 것)

---
현재 LB상 가장 높은 점수 기록하고 있으신
서일님의 토론글 노트북에서 보여준 검증전략은 다음과 같음.

0. 모델학습은 LightGBM 사용중
1. 10-Fold 검증 수행
2. Valid Set을 그대로 구축하는 방법도 있고, 혹은 그 중 꼭 필요하다고 생각되는 부분만 골라서 Valid로서 삼는 방법도 제안하고 있음
    1) 예를 들어서, 실제 LB에 제출해야하는 `AssentmentID`에 해당하는 해당하는 row만 Valid에 포함시키는 방법 고려하고 있음.
    2) 또한, 현재는 userID의 마지막 문항으로 검증셋이 구축되고 있는 상태인데
    3) 여기서 추가로 testID별 마지막 문항에 해당하는 row만 골라서 검증셋으로 구축하는 방법도 고려하고 있음

3. TrainSet도 그대로 구축하는 방법이 있는 반면, 그 중 꼭 필요하다고 판단되는 부분만 추출하여 Train하는 방법 역시 제안하고 있음
    1)  testID별 마지막 문항만 골라서 Train으로서 사용하는 방법
    2) userID별 마지막 문항만 골라서 Train으로서 사용하는 방법
    3) 실제 LB에 제출해야하는 `AssentmentID`에 해당하는 문제들로만 Train하는 방법
        - (이건 내가 추가하는 부분 : 위와 같은 방법이 있는 거라면, 실제 LB에 제출해야하는 `TestID`에 해당하는 문제들로만 Train하는 방법도 있겠네?)

> 사실 나는 `train 다쓰는게 당연히 더 좋은거 아니야?` 라고 생각하는 편인데 현재 Task에서는 row 수가 너무 많다보니..
위와 같이 주관적인 판단과 기준을 거쳐서 유의미한 row만 골라서 학습하는게 오히려
noise 를 없애줄 수 있는 효과가 있는걸까? and 와 그럼 이거 좋을 수도 있겠는데?라고 생각이 바뀌는 중임.

4. **결론적으로**, 서일님이 토론글에 올려줄 당시 가장 성능이 좋았던 (0.7995) 방법은
  - **TrainSet 구축시**에 `시험지마지막응답` 데이터만을 사용하였고
  - **ValidSet 구축시**에는 실제 LB에 제출해야하는 `AssentmentID`에 해당하는 해당하는 row만 Valid에 포함시킴.
  - 또한 **ValidSet 구축시** `userID`의 마지막 row, 즉 `사용자마지막응답` row들만 Valid에 포함시킴 (이건 사실 기존에도 적용하고 있는 방법)
<br>

나 역시도 그래서 여러 방법들 중 우선적으로 서일님과 같은 전략(위 4번)을 적용해보았고 그에 따른 결과는 다음과 같았음.(seed=42)
1) 이전의 검증전략을 통해서는 **Valid AUC : 약 0.795 -> LB AUC : 0.756** (차이가 약 0.04 정도)
2) 새로 바꾼 검증전략을 통해서는 **Valid AUC : 약 0.7556 -> LB AUC : 0.7512** (차이가 약 0.0044 정도니까 그 간극이 많이 줄어듦)
> 많은 실험들이 추가적으로 필요하겠지만.. 단편적인 실험결과는 위와 같이 좋아보였음

>여담)
  > 결국 LB AUC는 떨어졌잖아? 이게 뭐야? => ~들켰네...ㅠㅠ~ 가 아니라
      - 변명해보자면 실험환경이 똑같지가 않았어 ㅎㅎ..
      - `무지성 피쳐`가 2)에 2개 더 많은 상황이라 원래 성능이 더 안좋은 상태였고, 1)은 5Fold 2)는 10Fold를 썼음..
      - 그리고 설사 떨어졌다고 하더라도 나는 Valid Score 과대평가가 너무 심했는데, 그 간극을 엄청 많이 줄여냈다는 사실에 기뻤던 것이라 그리 부정적인 의미를 부여하지는 않았음!
<br>

## 추가적으로 끄적이고 정리하는 글..

먼저, **유의해야 할 사실**은 `나는 Valid Score의 과대평가가 매우 심한 상황이었음..`, 결국 이를 해결한 것 같아서(?) 기쁜 마음으로 지금 글을 쓰고는 있는 것이지만!
만약 자신의 Valid Score가 LB에 비해서 그리 과대평가되고 있는 상황이 아니라면 본인의 검증전략을 그대로 쓰고 이 글은 참고로만 봐도 괜찮을 것 같음!
<br>
끝으로, 여러가지 실험해보기에는 시간 소요가 많은 구조라서 이렇게 소수의 실험, 단편적인 결과밖에 못 보여준게 아쉽고..
위 내용들은 도움이 되긴 했지만 어쩌면 LGBM에서만 유의할 수 있는 방법일 수도 있기에.. 딥러닝에도 과연 효과를 볼 수 있을까? 의문이 드는 상황이긴 함.
  -  예를 들어서 RNN, Transformer 계열 모델들은 Input으로서 time정보들이 Sequential하게 들어가서 의미가 있다 보니 위와 같은 방법으로 접근하여 Train data를 줄일 필요가 전혀 없을 수도 있겠지...
  - 반면에 time 정보를 최대한 잃지 않으면서 유의미한 row만 골라 학습에 사용하는 방법론을 적용해볼 수 있다면, 이는 딥러닝에도 효과를 볼 수도 있겠지? (하지만 이것 역시 가설일뿐..)

- **그리고 나중에 혹시 시간이 된다면 (지금은 다른거 할게 많지만) 위와 같이 Valid 구축하면서 StratifiedKFold를 한번더 적용해보고 싶은 생각도 있음!!**

> 뭔가 이해 안되시는 부분 혹은 내가 잘못 적은 부분이 있다면 편하게 말해주세요~!

**Comment by arabae** [comment]
0.04에서 0.0044로 줄었다면 정말 의미있었던 실험인 것 같아😊 토론글의 노트북 내부까지는 못 봤는데 오빠 덕분에 빠르게 이해하고 넘어갑니다~

**Reply by MoonJaeHoon** [reply]
도움됐다니 기쁠 따름~
